hilti/toolchain/include/compiler/detail/cfg.h

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020-2023 by the Zeek Project. See LICENSE for details.
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <hilti/ast/all.h>
+#include <hilti/base/graph.h>
+
+namespace hilti {
+
+namespace node::tag {
+constexpr Tag MetaNode = 20000;
+constexpr Tag Start = 20001;
+constexpr Tag End = 20002;
+constexpr Tag Flow = 20003;
+constexpr Tag ScopeEnd = 20004;
+} // namespace node::tag
+
+namespace detail::cfg {
+struct MetaNode : Node {
+    // Some versions of GCC incorrectly diagnose maybe uninitialized members here.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpragmas"
+#pragma GCC diagnostic ignored "-Wunknown-warning-option"
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+    MetaNode(node::Tags node_tags) : Node(nullptr, node_tags, {}, {}) {}
+#pragma GCC diagnostic pop
+
+    uint64_t counter = instances++;
+    static uint64_t instances;
+    HILTI_NODE_0(MetaNode, override);
+};
+
+// A meta node for the start of a control flow.
+struct Start : MetaNode {
+    Start() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Start, MetaNode, final);
+};
+
+// A meta node for the end of a control flow.
+struct End : MetaNode {
+    End() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(End, MetaNode, final);
+};
+
+// A meta node joining or splitting control flow with no matching source statement.
+struct Flow : MetaNode {
+    Flow() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Flow, MetaNode, final);
+};
+
+// A meta node to signify end of a scope carrying the source range of that scope.
+struct ScopeEnd : MetaNode {
+    ScopeEnd(const Node* scope) : MetaNode(NodeTags), scope(scope) {
+        assert(scope); // Should always contain a valid scope.
+    }
+
+    HILTI_NODE_1(ScopeEnd, MetaNode, final);
+
+    const Node* scope;
+};
+
+struct Reachability {
+    std::unordered_set<Node*> in;
+    std::unordered_set<Node*> out;
+};
+
+struct Transfer {
+    std::unordered_set<Node*> use;
+    std::unordered_map<Node*, Node*> gen;
+    std::unordered_map<Node*, std::unordered_set<Node*>> kill;
```

Comment:
I'm not sure I fully understand `kill`. Example:

```
function void f(inout int<64> i) {
     i = 1;
     i = 2;
 }
```
![f](https://github.com/user-attachments/assets/d2c1a696-8434-4842-9013-add5d118cb80)

Looking at node 0, why is it killing `i: [i = 2;]`?

Reply:
FTR, we discussed this more offline, and the issue is that what this currently has as "kill sets" isn't really a kill set. We should instead align this more with the classic meaning, i.e., for each node in the graph track reachable inputs, and kill sets come then naturally. Having that data available in each node might also make it unnecessary to inspect the full graph to gather related information.